sdk/src/azure/core/az_mqtt5_rpc_client_hfsm.c

```diff
@@ -0,0 +1,691 @@
+  {
+    parent_state = root;
+  }
+  else
+  {
+    // Unknown state.
+    az_platform_critical_error();
+    parent_state = NULL;
+  }
+
+  return parent_state;
+}
+
+AZ_INLINE az_mqtt5_rpc_client_rsp_event_data _parse_response(
+    az_mqtt5_recv_data* recv_data,
+    az_mqtt5_property_binarydata* correlation_data,
+    az_mqtt5_property_stringpair* status,
+    az_mqtt5_property_stringpair* error_message,
+    az_mqtt5_property_string* content_type)
+{
+  // parse response
+  printf("Received response: %s\n", az_span_ptr(recv_data->payload));
+
+  az_mqtt5_rpc_client_rsp_event_data resp_data = { .response_payload = recv_data->payload,
+                                                   .parsing_failure = false,
+                                                   .status = AZ_MQTT5_RPC_STATUS_UNKNOWN,
+                                                   .error_message = AZ_SPAN_EMPTY,
+                                                   .content_type = AZ_SPAN_EMPTY,
+                                                   .correlation_id = AZ_SPAN_EMPTY };
+
+  // az_mqtt5_property_binarydata correlation_data;
+  if (az_result_failed(az_mqtt5_property_bag_read_binarydata(
+          recv_data->properties, AZ_MQTT5_PROPERTY_TYPE_CORRELATION_DATA, correlation_data)))
+  {
+    resp_data.parsing_failure = true;
+    resp_data.error_message
+        = AZ_SPAN_FROM_STR("Cannot process response message without CorrelationData");
+    return resp_data;
+  }
+  resp_data.correlation_id = az_mqtt5_property_get_binarydata(correlation_data);
+  printf("Processing response for: %s\n", az_span_ptr(resp_data.correlation_id));
+
+  // read the status of the response
+  // az_mqtt5_property_stringpair status;
+  if (az_result_failed(az_mqtt5_property_bag_find_stringpair(
+          recv_data->properties,
+          AZ_MQTT5_PROPERTY_TYPE_USER_PROPERTY,
+          AZ_SPAN_FROM_STR(AZ_MQTT5_RPC_STATUS_PROPERTY_NAME),
+          status)))
+  {
+    resp_data.parsing_failure = true;
+    resp_data.error_message = AZ_SPAN_FROM_STR("Response does not have the 'status' property.");
+    return resp_data;
+  }
+
+  // parse status
+  az_span status_str = az_mqtt5_property_stringpair_get_value(status);
+  if (az_result_failed((az_span_atoi32(status_str, (int32_t*)&resp_data.status))))
+  {
+    resp_data.parsing_failure = true;
+    resp_data.error_message = AZ_SPAN_FROM_STR("Status property contains invalid value.");
+    return resp_data;
+  }
+
+  if (resp_data.status < 200 || resp_data.status >= 300)
+  {
+    // read the error message if there is one
+    // az_mqtt5_property_stringpair error_message;
+    if (!az_result_failed(az_mqtt5_property_bag_find_stringpair(
+            recv_data->properties,
+            AZ_MQTT5_PROPERTY_TYPE_USER_PROPERTY,
+            AZ_SPAN_FROM_STR(AZ_MQTT5_RPC_STATUS_MESSAGE_PROPERTY_NAME),
+            error_message)))
+    {
+      resp_data.error_message = az_mqtt5_property_stringpair_get_value(error_message);
```

Comment:
I think this should always be checked, regardless of status.

Reply:
Also, maybe we need to have a status enum and response member that applications can use to decide if the status meaning is RPC was successful or failed.

Reply:
Will track this as an architectural question - for now the guidance we've received is that there should not be an error message when the status is successful